samplers/aws/xray/internal/reservoir.go

```diff
@@ -0,0 +1,112 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
+
+import (
+	"sync"
+	"time"
+)
+
+// reservoir represents a sampling statistics for a given rule and populate it's value from
+// the response getSamplingTargets API which sends information on sampling statistics real-time
+type reservoir struct {
+	// quota expiration timestamp
+	expiresAt time.Time
+
+	// quota assigned to client to consume per second
+	quota float64
+
+	// current balance of quota
+	quotaBalance float64
+
+	// total size of reservoir consumption per second
+	capacity float64
+
+	// quota refresh timestamp
+	refreshedAt time.Time
+
+	// polling interval for quota
+	interval time.Duration
+
+	// stores reservoir ticks
+	lastTick time.Time
+
+	mu *sync.RWMutex
```

Comment:
I think to avoid having this warning - https://stackoverflow.com/questions/37052757/how-to-get-around-assignment-copies-lock-value-to-tr-net-http-transport-contain

Reply:
If you work with a `*reservoir` instead of a `reservoir` (similar to [the linked answer](https://stackoverflow.com/a/37052960/3614233)), you should not have an issue with copying the lock. All of your method receivers are pointers so it seems like that might be the better approach, no?

Reply:
I think now that I remember the main motivation of making `reservoir` struct non pointer was to get a new address space when we deep copy the manifest object so don't think making `reservoir` pointer would help in this case. Also, since  we are doing deep copy I'm not sure if it's a good idea to keep this non pointer since if reservoir's `mu` gets updated we would be overriding stale lock values.

Reply:
I'm not sure this is true. we're applying read lock and that is also for a very brief period of time. This is how the code flow will look like. In below flow we're assuming mutex is a pointer.

Code applies read lock  -> Deep copy is performed (new copy has a read locked mutex) -> code release the lock (both copies are available for updates)

While we're doing a deep copy there won't be any updates anyways. so no need to worry about that even a very brief period of time. 

Yeah, I still don't think custom copy is needed here since there are already open sourced libraries available. I don't think deep copy is issue here anyways. Also, do you see any potential issues with mutex being a pointer? Is it not considered a good practice in golang?

Reply:
There is a reason govet warns about copying locks, and it is for this very reason. You make the copy and the lock now applies to 2 (or more) values.

What are you doing if you lock the original lock? You have locked both the original and the copy (or copies). Should they still be updated? According to their lock status they should not.

Should the original be locked if the copy is locked? As it is currently designed it will be.

When you make a copy of the original it needs to produce a value that has its own lock. If you can find a way to achieve this without a custom copy method that sounds good, but otherwise you need to make a custom copy method to ensure this.